bionic/task_state.py

```diff
@@ -396,28 +390,34 @@ def strip_state_for_subprocess(self, new_task_states_by_key=None):
             new_task_states_by_key = {}
 
         # All task keys should point to the same task state.
-        if self.task.keys[0] in new_task_states_by_key:
-            return new_task_states_by_key[self.task.keys[0]]
+        if self.task_keys[0] in new_task_states_by_key:
+            return new_task_states_by_key[self.task_keys[0]]
 
         # Let's make a copy of the task state.
         # This is not a deep copy so we'll avoid mutating any of the member variables.
         task_state = copy.copy(self)
 
         # Clear up memoized cache to avoid sending it through IPC.
         task_state._results_by_name = None
+        # Clear up fields not needed in subprocess, for computing or for cache lookup.
+        task_state.provider = None
+        task_state.case_key = None
+        task_state._provenance = None
 
         # Any non-persisted value will have to be recomputed again in the other
         # subprocess and isn't complete anymore.
-        if not task_state.provider.attrs.should_persist():
+        if not task_state.should_persist:
             task_state.is_complete = False
 
         if task_state.is_complete:
             task_state.dep_states = []
+            task_state.task = None
```

Comment:
It's getting even harder to reason about what states a `TaskState` can be in now....

Not a blocker for this PR, but something to keep an eye on.